kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/PaginatedScanFilesIteratorImpl.java

```diff
@@ -34,21 +46,68 @@ public PaginatedScanFilesIteratorImpl(
     this.pageSize = paginationContext.getPageSize();
   }
 
+  // Bug: we can't call hasNext() twice
   @Override
-  public Row getCurrentPageToken() {
-    return null;
+  public boolean hasNext() {
+    if (closed) {
+      throw new IllegalStateException("Can't call `hasNext` on a closed iterator.");
+    }
+    if (!nextBatch.isPresent()) {
+      prepareNext();
+    }
+    return nextBatch.isPresent();
   }
 
-  @Override
-  public boolean hasNext() {
-    return false;
+  private void prepareNext() {
+    if (numScanFilesReturned >= pageSize) {
+      return;
+    }
+    if (originalIterator.hasNext()) {
+      FilteredColumnarBatch batch = originalIterator.next();
+      String fileName = batch.getFilePath().get();
+      if (!fileName.equals(currentLogFilePath)) {
+        currentLogFilePath = fileName;
+        System.out.println("fileName " + fileName);
+        currentRowIdxInFile = 0; // row idx starts from 1
+      }
+      long numActiveAddFiles = batch.getPreComputedNumSelectedRows().get();
```

Comment:
`numActiveAddFiles` in what? in what context? probably `numActiveAddFilesInBatch`

but it's not just "active" -- we have also applied data skipping here, too, right? so perhaps `numSelectedAddFilesInBatch`

Reply:
also check that this `.get` exists, too